### ISSUE TITLE
Bug with SampleExtraction

### ISSUE THREAD

**Issue by remicres** [orfeotoolbox/pyotb#1]
It looks like there is a bug somewhere in `app.set_parameters` in `core.py`?

# How to reproduce

OTB 9
pyotb 2.0.1

## Works

```python
import otbApplication

app = otbApplication.Registry.CreateApplication("SampleExtraction")
app.SetParameterString("in", img_10m)
app.SetParameterString("vec", pos_a)
app.SetParameterString("field", "class")
app.SetParameterString("out", out)
app.ExecuteAndWriteOutput()
```

## Doesn't work

```python
import pyotb

pyotb.SampleExtraction({"in": img_10m, "vec": pos_a, "field": "class", "out": out})
```

The error:

```
Traceback (most recent call last):
  File "/home/otbuser/.local/lib/python3.10/site-packages/pyotb/core.py", line 743, in set_parameters
    self.__set_param(key, obj)
  File "/home/otbuser/.local/lib/python3.10/site-packages/pyotb/core.py", line 1027, in __set_param
    self.app.SetParameterValue(key, obj)
  File "/opt/otbtf/lib/otb/python/otbApplication.py", line 2898, in SetParameterValue
    raise TypeError("Expected a list for '%s' parameter in %s, got %s for %s" % (paramKey, self.GetName(), type(value), value))
TypeError: Expected a list for 'field' parameter in SampleExtraction, got <class 'str'> for class

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/code/part_2_sample_extraction_rf.py", line 36, in <module>
    pyotb.SampleExtraction({"in": img_10m, "vec": pos_a, "field": "class", "out": out})
  File "<string>", line 4, in __init__
  File "/home/otbuser/.local/lib/python3.10/site-packages/pyotb/core.py", line 603, in __init__
    self.set_parameters(*args, **kwargs)
  File "/home/otbuser/.local/lib/python3.10/site-packages/pyotb/core.py", line 745, in set_parameters
    raise RuntimeError(
RuntimeError: SampleExtraction: error before execution, while setting 'field' to 'class': Expected a list for 'field' parameter in SampleExtraction, got <class 'str'> for class)

```

**Comment by vijdz** [comment]
"Doesn't work", is that all ? No error message ?

**Comment by remicres** [comment]
Same for `TrainVectorClassifier`

```python
# doesn't work
pyotb.TrainVectorClassifier(
  io_vd=in_a,
  valid_vd=in_b,
  feat=feats,
  cfield="class",
  classifier="rf",
  io_out=out
)

# works
app = otbApplication.Registry.CreateApplication("TrainVectorClassifier")
app.SetParameterStringList("io.vd", [in_a])
app.SetParameterStringList("valid.vd", [in_b])
app.SetParameterStringList("feat", feats)
app.SetParameterString("cfield", "class")
app.SetParameterString("classifier", "rf")
app.SetParameterString("io.out", out)
app.ExecuteAndWriteOutput()
```


```
2024-04-03 13:32:10.490090: I tensorflow/core/util/port.cc:111] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-04-03 13:32:10 (INFO) [pyotb] Successfully loaded 116 OTB applications
2024-04-03 13:32:10 (INFO) [pyotb] TrainVectorClassifier: argument for parameter "io.vd" was converted to list
2024-04-03 13:32:10 (INFO) [pyotb] TrainVectorClassifier: argument for parameter "valid.vd" was converted to list
Traceback (most recent call last):
  File "/home/otbuser/.local/lib/python3.10/site-packages/pyotb/core.py", line 743, in set_parameters
    self.__set_param(key, obj)
  File "/home/otbuser/.local/lib/python3.10/site-packages/pyotb/core.py", line 1043, in __set_param
    raise TypeError(
TypeError: TrainVectorClassifier: wrong parameter type (<class 'list'>) for 'feat'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/code/part_2_train_rf.py", line 11, in <module>
    pyotb.TrainVectorClassifier(
  File "<string>", line 4, in __init__
  File "/home/otbuser/.local/lib/python3.10/site-packages/pyotb/core.py", line 603, in __init__
    self.set_parameters(*args, **kwargs)
  File "/home/otbuser/.local/lib/python3.10/site-packages/pyotb/core.py", line 745, in set_parameters
    raise RuntimeError(
RuntimeError: TrainVectorClassifier: error before execution, while setting 'feat' to '['value_0', 'value_1', 'value_2', 'value_3']': TrainVectorClassifier: wrong parameter type (<class 'list'>) for 'feat')

```

**Comment by remicres** [comment]
> "Doesn't work", is that all ? No error message ?

@vidlb edit: I have updated the description

**Comment by vijdz** [comment]
Can you reproduce with OTB 8 ?  
If it works in OTB 8 and not in OTB 9, it's not a pyotb bug, I guess ParameterType enums may have changed.

**Comment by vijdz** [comment]
Looking at your example, it has to be linked to `SetParameterValue`, because we use this and not `SetParameterString`

**Comment by vijdz** [comment]
Can you try with otbApplication and ` app.SetParameterValue("field", "class")` ?  I'm pretty sure that's the problem
So maybe we'll have to use specific funcs for specific types once again. Gonna be a nightmare.  
SetParameterValue was to easy, it couldn't last long

**Comment by remicres** [comment]
Indeed, I have the same error when I try with `SetParameterValue`. This was working in OTB 8 ?

**Comment by vijdz** [comment]
Not sure since we don't have those apps in the tests, but I guess you would have found out if not.  
It's probably affecting other apps

**Comment by vijdz** [comment]
I just ran pytest with OTB 9 and it seems to work except for the summarize tests : the float precision error is back, I thought that was merged ?

**Comment by vijdz** [comment]
> This was working in OTB 8 ?

Just tried and it's the same error. So we missed that one type "field" for which SetParameterValue doesn't work. I can patch this pretty fast but we'll need more tests for those apps now.

**Comment by remicres** [comment]
It looks like the expected parameter is something exclusive for vector data

```
'ParameterType_Field'
```

**Comment by vijdz** [comment]
@remicres can you test with that branch ? https://gitlab.orfeo-toolbox.org/nicolasnn/pyotb/-/tree/bug-param-type-field
I pushed a dev tag for a pre-release, but I forget the pipeline process is completely broken right now due to the docker image issue...

**Comment by vijdz** [comment]
So what now, should we move the whole thing to github, gitlab or forgemia ? With our own docker image for testing

**Comment by remicres** [comment]
> So what now, should we move the whole thing to github, gitlab or forgemia ? With our own docker image for testing

+1

I made a fork on forgemia

**Comment by vijdz** [comment]
@remicres can you test this with 2.0.2 and close if it's working ? Also we'll need to change this mirror upstream from otb git to forgemia, I don't have the right permissions here to do it.

**Comment by remicres** [comment]
> we'll need to change this mirror upstream from otb git to forgemia, I don't have the right permissions here to do it.

Done.
Let's continue the other thread on forgemia

**Comment by vijdz** [comment]
And same regarding nicolasnn/pyotb, I'm not owner there, what should we ? Archive, remove ?

**Comment by remicres** [comment]
archive when all the migration is complete yes

**Comment by vijdz** [comment]
> archive when all the migration is complete yes

isn't it already ?

**Comment by remicres** [comment]
Solved in `2.0.2`, thanks a lot @vidlb
